---
title: "R Notebook"
output: html_notebook
---
```{r error=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
library(Seurat)
library(patchwork)
library(plyr)
library(magrittr)
library(tidyr)
library(ggplot2)
library(ggthemes)
library(fs)
library(roxygen2)
library(utils)
library(fs)
library(knitr)
library(purrr)
library(reshape)
library(ggplot2)
library(matlab)
library(matrixStats)
library(cowplot)
library(openxlsx)
library(readxl)
library(harmony)
library(SoupX)
library(ggpubr)
library(pheatmap)
library(scales)
library(gridExtra)
library(stargazer)
library(data.table)
library(rmarkdown)
library(monocle)
library(SeuratWrappers)
library(monocle) 
library(pheatmap)
library(gprofiler2)
library(stringr)
library(RColorBrewer)
library(moonBook)
library(ggforce)
library(yarrr)

 basic.marker1 <- c( "Lepr", "Cemip", "Mag", "Pdgfra", "Gfap", "Slc1a2", "Ptprc", "Syt1", "Nrg1", "Tmem232", "Cux2", 'S100a8','Gad1', 'Slc17a7', 'Flt1', "Mbp", 'Tmem119', 'P2ry12', 'Trem2', 'Itgam', 'Itgax', 'Mecom', 'Pdgfrb', 'Acta2', 'Apoe', 'Top2a', 'Enpp6', 'Cspg4','Cd14')
```

# Function to save figures
```{r}
fig.path <- '/Users/brakeam/OneDrive - National Institutes of Health/JP & Alexis/20240419_newDTA/3_plots.output/'

fig.path.png <- '/Users/brakeam/OneDrive - National Institutes of Health/JP & Alexis/20240419_newDTA/3_plots.output/1_PNG/AST'
fig.path.png <- '/Users/brakeam/OneDrive - National Institutes of Health/JP & Alexis/20240419_newDTA/3_plots.output/2_PDF/AST'

save.fig <- function(fig.path,p.name) {
ggsave(filename = paste0(fig.path.pdf, 
                         p.name,'_',
                         'GROUP.',resolution, "_",
                         format(Sys.Date(), "%Y%m%d"),
                         '.PDF'), 
       limitsize = F) 
  
  ggsave(filename = paste0(fig.path.png,
                         p.name,'_',
                         'GROUP.',resolution, "_",
                         format(Sys.Date(), "%Y%m%d"),
                         '.png'), 
       limitsize = F) 
}
  
```

# Functions for dotplots and feature plots split by a category with titles
```{r}

FPlist <- function(genes){
    FeaturePlot(IMM_T2T3, genes ,ncol = n.col,cols = c('grey90',paste0('#',c("4FB3AA","e9d8a6","ee9b00","ae2012"))),pt.size = 0.05, order = order.input) & NoAxes() 
} #function to plot markers for each cluster 


DPlist <- function(genes){
   DotPlot(IMM_T2T3,features=genes,group.by = 'RNA_snn_res.0.6', dot.scale = 6,cluster.idents = F,col.max = 1.5, col.min = -1.5)+
    theme(axis.text.x = element_text(angle = 0, hjust = 1)) +
  scale_colour_gradient2(low = "#046A83", mid = "white",high = '#D26952',midpoint = 0) + theme_light()  + 
  rotate_x_text(60) +
    theme( plot.margin = margin(b = 0, r = 20),
          legend.position = 'right',
          axis.title.x = element_text(size = 0),
          axis.title.y = element_text(size = 0)) +
    geom_point(aes(size = pct.exp), alpha = 0.25, shape = 21, color="black", stroke = 0.75)+
  FontSize(x.text = 12,y.text = 12) + coord_flip()
} #function to plot markers for each cluster 
```

# Read data file 
```{r}
IMM_T2T3 <-readRDS('/Users/brakeam/OneDrive - National Institutes of Health/JP & Alexis/20240419_newDTA/1_data.input/1_DTA/Cleaned_T2T3only/IMM_T2T3_20240726.RDS')
```

# UPDATE INDEX LABELS
```{r}
index_labels <- readxl::read_excel('/Users/brakeam/Documents/Popko/snRNA-seq/Popko_snRNAseq_Analysis/data/index_labels_b1_b2_b3_b4.xlsx', sheet = 'IndexLabels')


order <- match(dataSeurat@meta.data$IL01_uniqueID, index_labels$IL01_uniqueID)
dataSeurat@meta.data$Test <- as.character(index_labels[order,]$Test)
dataSeurat@meta.data$IL02_species <- as.character(index_labels[order,]$IL02_species)
dataSeurat@meta.data$IL03_source.1 <- as.character(index_labels[order,]$IL03_source.1)
dataSeurat@meta.data$IL03_source.2 <- as.character(index_labels[order,]$IL03_source.2)
dataSeurat@meta.data$IL04_sex <- as.character(index_labels[order,]$IL04_sex)
dataSeurat@meta.data$IL05_ageDays <- as.character(index_labels[order,]$IL05_ageDays)
dataSeurat@meta.data$IL05_ageDays.1 <- as.character(index_labels[order,]$IL05_ageDays.1)
dataSeurat@meta.data$IL05_ageDays.2 <- as.character(index_labels[order,]$IL05_ageDays.2)
dataSeurat@meta.data$IL06_tissue.1 <- as.character(index_labels[order,]$IL06_tissue.1)
dataSeurat@meta.data$IL06_tissue.2 <- as.character(index_labels[order,]$IL06_tissue.2)
dataSeurat@meta.data$IL06_tissue.3 <- as.character(index_labels[order,]$IL06_tissue.3)
dataSeurat@meta.data$IL07_location <- as.character(index_labels[order,]$IL07_location)
dataSeurat@meta.data$IL08_condition <- as.character(index_labels[order,]$IL08_condition)
dataSeurat@meta.data$IL08_condition.1 <- as.character(index_labels[order,]$IL08_condition.1)
dataSeurat@meta.data$IL08_condition.2 <- as.character(index_labels[order,]$IL08_condition.2)
dataSeurat@meta.data$IL09_ilumina <- as.character(index_labels[order,]$IL09_ilumina)
dataSeurat@meta.data$IL10_chemistry <- as.character(index_labels[order,]$IL10_chemistry)
dataSeurat@meta.data$IL11_batch <- as.character(index_labels[order,]$IL11_batch)
dataSeurat@meta.data$IL12_LMinDays <- as.character(index_labels[order,]$IL12_LMinDays)
dataSeurat@meta.data$IL13_LmaxDays <- as.character(index_labels[order,]$IL13_LmaxDays)
dataSeurat@meta.data$IL14_dataset <- as.character(index_labels[order,]$IL14_dataset)
dataSeurat@meta.data$IL16_note <- as.character(index_labels[order,]$IL16_note)

dataSeurat$IL08_condition.5 <- ""
dataSeurat$IL08_condition.5 <- paste0(dataSeurat$IL08_condition.2,"_",dataSeurat$IL08_condition)

dataSeurat$IL08_condition.6 <- ''
dataSeurat$IL08_condition.6[which(dataSeurat$IL08_condition %in% c('3w'))] <-  'T1_Onset'
dataSeurat$IL08_condition.6[which(dataSeurat$IL08_condition %in% c('5w','~6.28w'))] <-  'T2_Peak'
dataSeurat$IL08_condition.6[which(dataSeurat$IL08_condition %in% c('17w'))] <-  'T3_Recovery'
dataSeurat$IL08_condition.6[which(dataSeurat$IL08_condition %in% c('~45w'))] <-  'T4_LateOnset'
dataSeurat$IL08_condition.6 <- factor(dataSeurat$IL08_condition.6)



dataSeurat$IL08_condition.5 <- factor(dataSeurat$IL08_condition.5, levels =c('ctrl_3w','DTA_3w','ctrl_5w', 'DTA_~6.28w','ctrl_17w', 'DTA_17w','ctrl_~45w','DTA_~45w'))

dataSeurat$IL08_condition.7 <- ""
dataSeurat$IL08_condition.7 <- paste0(dataSeurat$IL08_condition.6,"_",dataSeurat$IL08_condition.2)
dataSeurat$IL08_condition.7 <- factor(dataSeurat$IL08_condition.7,levels =c('T1_Onset_ctrl','T2_Peak_ctrl','T3_Recovery_ctrl', 'T4_LateOnset_ctrl','T1_Onset_DTA','T2_Peak_DTA','T3_Recovery_DTA', 'T4_LateOnset_DTA'))

dataSeurat$IL08_condition.8 <- ""
dataSeurat$IL08_condition.8 <- paste0(dataSeurat$IL08_condition.7,"_",dataSeurat$IL06_tissue.1)
dataSeurat$IL08_condition.8 <- factor(dataSeurat$IL08_condition.8,levels =c('T1_Onset_ctrl_GM','T2_Peak_ctrl_GM','T3_Recovery_ctrl_GM', 'T4_LateOnset_ctrl_GM','T1_Onset_DTA_GM','T2_Peak_DTA_GM','T3_Recovery_DTA_GM', 'T4_LateOnset_DTA_GM','T1_Onset_ctrl_WM','T2_Peak_ctrl_WM','T3_Recovery_ctrl_WM', 'T4_LateOnset_ctrl_WM','T1_Onset_DTA_WM','T2_Peak_DTA_WM','T3_Recovery_DTA_WM', 'T4_LateOnset_DTA_WM'))
```

# Map my cells 
```{r set working directory}
# Uncomment line below and replace bracketed text with path to downloaded files.
setwd("/Users/brakeam/OneDrive - National Institutes of Health/JP & Alexis/20240419_newDTA/6_AnalysisOutput/IMM")

```
  
#### Load the relevant libraries
This workbook uses the libraries *anndata* and *Seurat*
```{r load libraries, warnings=FALSE}
# Install Seurat and anndata if needed
list.of.packages <- c("anndata")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)>0) install.packages(new.packages)

# Load Seurat and anndata
suppressPackageStartupMessages({
  library(Seurat)      # For reading droplet data sets and visualizing/comparing results
  library(anndata)     # For writing h5ad files 
})
options(stringsAsFactors=FALSE)

# Citing R libraries
# citation("Seurat") # Note that the citation for any R library can be pulled up using the citation command. We encourage citation of R libraries as appropriate.
```
  
With the above files in your current working directory and the above libraries loaded, the use case below can now be run.  
  
### Analysis of a library
  
**(If you already have an .h5ad file ready for upload, skip to step 4.)**
  
A common use case for analysis of single cell/nucleus transcriptomics is to collect data from one (or more) ports of a droplet-based scRNA-seq run.  After some QC steps, these are then clustered for defining cell types. This section describes how to start from a such a droplet-based sequencing run, transfer labels from mouse "10x scRNA-seq whole brain" data from the Allen Institute onto these cells using MapMyCells, and then visualize the results in a UMAP.  
  
#### 3. Output to h5ad format

Now let's output the QC'ed data matrix into an h5ad file and output it to the current directory for upload to MapMyCells. Note that in anndata data structure the genes are saved as columns rather than genes so we need to transpose the matrix first.
  
```{r output library anndata, eval=FALSE}
IMM_T2T3_mat <- IMM_T2T3@assays[["RNA"]]@counts
dataQCt = Matrix::t(IMM_T2T3_mat)

# Convert to anndata format
ad <- AnnData(
  X = dataQCt,
  obs = data.frame(group = rownames(dataQCt), row.names = rownames(dataQCt)),
  var = data.frame(type = colnames(dataQCt), row.names = colnames(dataQCt))
)

# Write to compressed h5ad file
write_h5ad(ad,'IMM_T2T3.h5ad',compression='gzip')

# Check file size. File MUST be <500MB to upload for MapMyCells
print(paste("Size in MB:",round(file.size("IMM_T2T3.h5ad")/2^20)))
``` 
  
#### 4. Assign cell types using MapMyCells 
  
These next steps are performed OUTSIDE of R in the MapMyCells web application.  

The steps to MapMyCells are as follows:

1. Go to (https://knowledge.brain-map.org/mapmycells/process/).  
2. (Optional) Log in to MapMyCells.  
3. Upload 'droplet_library.h5ad' to the site via the file system or drag and drop (Step 1).  
4. Choose "10x Whole Mouse Brain (CCN20230722)" as the "Reference Taxonomy" (Step 2).  
5. Choose the desired "Mapping Algorithm" (in this case "Hierarchical Mapping").  
6. Click "Start" and wait ~5 minutes. (Optional) You may have a panel on the left that says "Map Results" where you can also wait for your run to finish.  
7. When the mapping is complete, you will have an option to download the "tar" file with mapping results.  If your browser is preventing popups, search for small folder icon to the right URL address bar to enable downloads.  
8. Unzip this file, which will contain three files: "validation_log.txt", "[NUMBER].json", and "[NUMBER].csv".  [NUMBER].csv contains the mapping results, which you need. The validation log will give you information about the the run itself and the json file will give you extra information about the mapping (you can ignore both of these files if the run completes successfully).  
9. Copy [NUMBER].csv to your current working directory and rename it "droplet_library_mapping.csv").  
10. You can now go back to R and continue the script below.  
  
#### 5. Read mapping results into R
  
Let's now look at the output results from the hierarchical clustering algorithm.  We can read this into R using read.csv, but note that the first four lines contain metadata that need to be skipped.  
  
```{r read library mapping results}
mapping <- read.csv("/Users/brakeam/Library/CloudStorage/OneDrive-NationalInstitutesofHealth/JP & Alexis/20240419_newDTA/6_AnalysisOutput/IMM/MapMyCells/IMM_T2T3_10xWholeMouseBrain(CCN20230722)_HierarchicalMapping_UTC_1722307615263/IMM_T2T3_10xWholeMouseBrain(CCN20230722)_HierarchicalMapping_UTC_1722307615263.csv",comment.char="#")
head(data.frame(mapping))
```   
  
MapMyCells maps input cells to the taxonomy at four increasing levels of resolution from coarsest class, to intermediate subclass, and supertype, and finest cluster. In the mouse whole brain taxonomy there are 32 classes, 306 subclasses, 1,045 supertypes and 5,200 clusters.   
  
The file consists of the following columns:

* *cell_id* = the cell identifiers for your cells included in the initial h5ad files  
* *class_label* = the unique identifier for the best mapping class  
* *class_name* = the name of the best mapping class  
* *class_bootstrapping_probability* = fraction of bootstrapping runs that the cell mapped to the above class (higher numbers are better, with 1 being the best)  
and similar fields for subclass, supertype, and cluster.  For finest level *cluster* there is an additional field  
* *cluster_alias* = another unique identifier for the cluster  
  
#### 6. Review top classes
  
As this library was selected from a dissection of primary motor cortex (MOp), we expect the majority of cells to map to cell types found in MOp.  Let's check!  

Now let's visualize the mapping results.  We will do this by saving the data in a Seurat object with (modified) mapping results as metadata, running the standard pipeline for creating a UMAP in Seurat, and then color-coding each cell.  
  
```{r define modified mapping, warning=FALSE}
# Assign rare classes and subclasses as "other"
mapping$class_new <- mapping$class_name
 mapping$class_new[!is.element(mapping$class_name,names(head(-sort(-table(mapping$class_name)),5)))] = "other"
 mapping$class_new[which(mapping$class_bootstrapping_probability <.99)] = "Probability < 0.99"

mapping$subclass_new <- mapping$subclass_name
mapping$subclass_new[!is.element(mapping$subclass_name,names(head(-sort(-table(mapping$subclass_name)),20)))] = "other"
mapping$subclass_new[which(mapping$subclass_bootstrapping_probability <.95)] = "Probability < 0.95"

mapping$supertype_new <- mapping$supertype_name
mapping$supertype_new[!is.element(mapping$supertype_name,names(head(-sort(-table(mapping$supertype_name)),6)))] = "other"
mapping$supertype_new[which(mapping$supertype_bootstrapping_probability <.95)] = "Probability < 0.95"

# Put row.names as data colnames and the order to match the data
rownames(mapping) <- mapping$cell_id
mapping <- mapping[colnames(IMM_T2T3_mat),]

# Create the Seurat object
dataSeurat <- IMM_T2T3
dataSeurat@meta.data <- cbind(IMM_T2T3@meta.data,mapping)
```   
 
```{r view library top classes}
# View the top classes
data.frame(Cell_counts=head(sort(table(mapping$supertype_new),decreasing=T),20))

library(dplyr)
dataSeurat@meta.data %>%
  group_by(RNA_snn_res.0.6, subclass_new) %>%
  dplyr::summarize(count=n())%>%
  filter(RNA_snn_res.0.6 == 5) 

clust <- dataSeurat@meta.data[which(dataSeurat$RNA_snn_res.0.6 %in% c(8)),]
clust.table <- data.frame(Cell_counts=sort(table(c11$class_new),decreasing=T))
hist(c11$subclass_bootstrapping_probability) 

mean(clust.table$Cell_counts.Freq)
``` 

Now let's make the plot for classes!  
  
```{r plot the umap classes, fig.height=8, fig.width=9}
DimPlot(dataSeurat, group.by="RNA_snn_res.0.6", split.by="subclass_new", label=TRUE, repel = T, ncol = 3, pt.size =.5) + NoAxes() 
DimPlot(dataSeurat, split.by="RNA_snn_res.0.6", group.by="subclass_new", label=F, repel = T, ncol = 3, pt.size =.1) + NoAxes() 

DimPlot(dataSeurat, group.by="subclass_new", split.by="class_new", label=F, repel = T, ncol = 3, pt.size =.5,cols = c('firebrick','orange','blue','yellow','magenta','seagreen4','mediumseagreen','darkviolet','darkgoldenrod1','dodgerblue4','red','lightsteelblue2','pink2','darkmagenta','darkcyan','grey75','black')) + NoAxes() 

DimPlot(dataSeurat, group.by="subclass_new", split.by="IL08_condition.7", label=F, repel = T, ncol = 2, pt.size =.3, cols = c('firebrick','mediumseagreen','darkgoldenrod1','dodgerblue4','lightsteelblue2','pink2','darkmagenta','darkcyan','grey75','black')) + NoAxes() 

FeaturePlot(dataSeurat,features = c('subclass_bootstrapping_probability','class_bootstrapping_probability'),cols =c('grey80',"#4FB3AA","#e9d8a6","#ee9b00","#ae2012"),blend= T) 


FeaturePlot(dataSeurat,features = c('class_bootstrapping_probability'),cols = c("#ae2012",'grey90'),blend= F, pt.size = 0.6) &NoAxes() | FeaturePlot(dataSeurat,features = c('subclass_bootstrapping_probability'),cols = c("#ae2012",'grey90'),blend= F, pt.size = 0.6) &NoAxes()
FeaturePlot(dataSeurat,features = c('supertype_bootstrapping_probability'),cols = c("#ae2012",'grey90'),blend= F, pt.size = 0.6) &NoAxes()
```   

```{r fig.height=3, fig.width=9}
DimPlot(dataSeurat, split.by="IL08_condition.7", group.by="subclass_new", label=F, repel = T, ncol = 4, pt.size =.25) + NoAxes() 
```

Here the data has not been clustered but rather the class assignments are assigned to the data in the UMAP. The alignment suggests that the label transfer works well.  It's worth noting that Seurat produces different UMAP configurations in different R environments, so your plot may not look exactly like this.  
  
Now let's make the plot where we color-code the same UMAP by subclass.  
  
```{r plot the umap subclasses, fig.height=6, fig.width=16}
DimPlot(dataSeurat, reduction = "umap", group.by="subclass_new", split.by = 'IL08_condition.7',label=F, repel  = T,pt.size = .5) 

table(dataSeurat$subclass_new)
```

```{r plot the umap subclasses, fig.height=2, fig.width=3}
hist(dataSeurat$class_bootstrapping_probability) 
hist(dataSeurat$subclass_bootstrapping_probability)

table(dataSeurat$subclass_bootstrapping_probability)
``` 
  
Once again, the subclasses shown largely segregate from one another without the need to apply clustering, suggesting this mapping works well even at higher resolutions. Scripts such as this can be used for visualizing and comparing mapping results.  
  
To output the session information we write the command, which is useful for reproducibility, especially for more complex scripts.  
  
```{r sessionInfo}
sessionInfo()
```
  
# Plot markers for the classes and subclasses identified 
# use both subclass.markers.combo and subclass.tf.markers.combo
  
```{r}
OBIMN.GABA.45 <- unique(dput(scan(text="2610307P16Rik,Dlx1,Zeb2,Pax6,Dlx2,Zeb2,Bcl11b,Meis2,Nfix", what=",", sep = ',')))
CB.Granule.GLUT.314 <- unique(dput(scan(text="Gabra6,Ror1,Pax6,Neurod2,Etv1", what=",", sep = ',')))
ASTRO.NT.NN.318 <- unique(dput(scan(text="Aqp4,Cd38,Agt,Itih3,Rfx4,Dbx2,Prdm16,Nr2f1", what="", sep = ',')))
Microglia.NN.318 <- unique(dput(scan(text="C1qa,Tmem119,Fli1,Sall1", what=",", sep = ','))) 
BAM.NN.335 <- unique(dput(scan(text="Mrc1,Maf,Fli1,Etv1,Runx1,Trps1", what=",", sep = ','))) 
DC.NN.337 <- unique(dput(scan(text="Cd74,Runx1,Batf3", what="", sep = ','))) 
Lymphoid.NN.338 <- unique(dput(scan(text="Cd52,Gimap3,Ikzf1,Ikzf3", what=",", sep = ','))) 

# make a list with all of them 
Allen.marker.list <- list(OBIMN.GABA.45,CB.Granule.GLUT.314,ASTRO.NT.NN.318,Microglia.NN.318,BAM.NN.335,DC.NN.337,Lymphoid.NN.338) 
  names(Allen.marker.list) <-  dput(scan(text="OBIMN.GABA.45,CB.Granule.GLUT.314,ASTRO.NT.NN.318,Microglia.NN.318,BAM.NN.335,DC.NN.337,Lymphoid.NN.338", what="", sep = ','))
  
genes <- plyr::ldply(Allen.marker.list,cbind)
colnames(genes) <- c('subclass','Gene') 
genes <- genes[which(isUnique(genes$Gene) == T),] # subset the df for only unique genes
rownames(genes) <- genes$Gene

Allen.marker.list.unique <- split(genes,genes$subclass)
Allen.marker.list.unique <- lapply(Allen.marker.list.unique, rownames)
```

## Plot allen markers from mapmycells top subclasses
```{r fig.height=14, fig.width=8}
order.input = T
n.col = 4
fp <- lapply(seuratobj = dataSeurat, X = Allen.marker.list.unique, FUN = FPlist)
rel.height <- ceil(as.numeric(lapply(Allen.marker.list.unique, length))/n.col) 

plot_grid(plotlist = fp, ncol = 1, rel_heights = rel.height, labels = paste0(names(fp)), label_colour = '#ae2012', vjust = 1.05, label_size = 16, scale = 0.95) #merge plots for all clusters into one big figure for export
```

```{r fig.height=8, fig.width=12}


dp <- lapply(seuratobj = dataSeurat, X = Allen.marker.list.unique, FUN = DPlist, resolution = 'RNA_snn_res.0.6' )
rel.height <- as.numeric(lapply(Allen.marker.list.unique, length)) #take the expected number of rows and based on length of markers for each cluster and the number of columns
 plot_grid(plotlist = dp, ncol = 3, labels = paste0(names(dp)), label_colour = '#ae2012', vjust = 1.0, label_size = 16, scale = 0.85, axis = 'l') #merge plots for all clusters into one big figure for export

 
dp <- lapply(seuratobj = dataSeurat, X = Allen.marker.list.unique, FUN = DPlist, resolution = 'class_new' )
rel.height <- as.numeric(lapply(Allen.marker.list.unique, length)) #take the expected number of rows and based on length of markers for each cluster and the number of columns
 plot_grid(plotlist = dp, ncol = 3, labels = paste0(names(dp)), label_colour = '#ae2012', vjust = 1.0, label_size = 16, scale = 0.85, axis = 'l') #merge plots for all clusters into one big figure for export
 
 
dp <-lapply(seuratobj = dataSeurat, X = Allen.marker.list.unique, FUN = DPlist, resolution = 'subclass_new' )
rel.height <- as.numeric(lapply(Allen.marker.list.unique, length)) #take the expected number of rows and based on length of markers for each cluster and the number of columns
 plot_grid(plotlist = dp, ncol = 3, labels = paste0(names(dp)), label_colour = '#ae2012', vjust = 1.0, label_size = 16, scale = 0.85, axis = 'l') #merge plots for all clusters into one big figure for export
```

```{r fig.height=7, fig.width=12}
DPlist(seuratobj = dataSeurat ,genes = genes$Gene, resolution = 'RNA_snn_res.0.6') | DPlist(seuratobj = dataSeurat ,genes = genes$Gene, resolution = 'class_new')  | DPlist(seuratobj = dataSeurat ,genes = genes$Gene, resolution = 'subclass_new') 
```
